File: leetcode_oj/right_side_view.py
'''
Given a binary tree, imagine yourself standing on the right side of it, return the values of the nodes you can see ordered from top to bottom.

For example:
Given the following binary tree,
   1            <---
 /   \
2     3         <---
 \     \
  5     4       <---
You should return [1, 3, 4].
'''
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x, l, r):
        self.val = x
        self.left = l
        self.right = r

class Solution:
    # @param {TreeNode} root
    # @return {integer[]}
	def rightSideView(self, root):
		if not root:
			return list()
		nodes_by_level = {0 : [root.val]}
		self.helper(root, nodes_by_level, 1)
		node_on_rightside = list()
		for level in nodes_by_level:
			if nodes_by_level[level]:
				node_on_rightside.append(nodes_by_level[level].pop())
		return node_on_rightside
		
	def helper(self, node, nodes_by_level, level):
		if level not in nodes_by_level:
			nodes_by_level[level] = list()
		if node.left:
			nodes_by_level[level].append(node.left.val)	
			self.helper(node.left, nodes_by_level, level + 1)
		if node.right:
			nodes_by_level[level].append(node.right.val)
			self.helper(node.right, nodes_by_level, level + 1)
		

s = Solution()
n7 = TreeNode(7, None, None)
n4 = TreeNode(4, n7, None)
n5 = TreeNode(5, None, None)
n2 = TreeNode(2, n4, None)
n3 = TreeNode(3, n5, None)
n1 = TreeNode(1, n2, n3)
print s.rightSideView(n1)

File: leetcode_oj/rightSideView.cpp
#include <iostream>
#include <vector>
using namespace std;
/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
	public:
    vector<int> rightSideView(TreeNode *root) {
        vector<int> result;
        rightView(root, result, 0);
        return result;
    }

    void rightView(TreeNode *curr, vector<int>& result, int currDepth){
        if(!curr){
            return;
        }
		/*
		cout << "Before appending" << "\n";
		cout << "node: " << curr->val << " currDepth: " << currDepth << "\n";
		copy(result.begin(), result.end(), ostream_iterator<int>(cout, " "));
		cout << "\n";
		cout << "\n";
		*/
        if(currDepth == result.size()){
            result.push_back(curr->val);
        }

        rightView(curr->right, result, currDepth + 1);
		/*
		cout << "After appending" << "\n";
		cout << "node: " << curr->val << " currDepth: " << currDepth << "\n";
		cout << "\n";
		cout << "\n";
		copy(result.begin(), result.end(), ostream_iterator<int>(cout, " "));
		*/
        rightView(curr->left, result, currDepth + 1);

    }
};
int main() {
	Solution sol;
	TreeNode root(1);
	TreeNode t2(2), t3(3), t4(4), t5(5);

	root.left = &t2; root.right = &t3;
	t2.left = &t4; t2.right = &t5;

	vector<int> result = sol.rightSideView(&root);
	copy(result.begin(), result.end(), ostream_iterator<int>(cout, " "));
	cout << "\n";
	return 0;
}
